cpp_api/objects/equation.cpp

```diff
@@ -2,355 +2,333 @@
     {
         IodeExceptionInvalidArguments error("Cannot extract Equation", "Equation position must be in range [0, " + 
             std::to_string(kdb->k_nb - 1) + "])");
         error.add_argument("equation position", std::to_string(pos));
         throw error;
     }
+
     // Note: KEVAL allocate a new pointer EQ*
-    c_equation = KEVAL(kdb, pos);
-    endo = std::string(KONAME(kdb, pos));
+    EQ* c_eq = KEVAL(kdb, pos);
+    copy_from_EQ_obj(c_eq);
+    E_free(c_eq);
+    // re-compute CLEC
+    char* name = KONAME(kdb, pos);
+    this->clec = L_solve(this->lec, name);
 }
 
 Equation::Equation(const std::string& name, KDB* kdb)
 {
-    if (!kdb) kdb = K_WS[I_EQUATIONS];
+    if (!kdb) 
+        kdb = K_WS[I_EQUATIONS];
+
     int pos = K_find(kdb, to_char_array(name));
-    if (pos < 0) throw IodeExceptionFunction("Cannot extract Equation", "Equation with name " + name + " does not exist.");
+    if (pos < 0) 
+        throw IodeExceptionFunction("Cannot extract Equation", "Equation with name " + name + " does not exist.");
+    
     // Note: KEVAL allocate a new pointer EQ*
-    c_equation = KEVAL(kdb, pos);
-    endo = name;
+    EQ* c_eq = KEVAL(kdb, pos);
+    copy_from_EQ_obj(c_eq);
+    E_free(c_eq);
+    // re-compute CLEC
+    this->clec = L_solve(this->lec, to_char_array(name));
+}
+
+Equation::Equation(const std::string& name, const std::string& lec, const int method, const std::string& from, const std::string& to, 
+    const std::string& comment, const std::string& instruments, const std::string& block, const bool date)
+{
+    this->lec = NULL, 
+    this->clec = NULL, 
+    this->cmt = NULL, 
+    this->instr = NULL, 
+    this->blk = NULL;
+    this->date = 0L;
+    this->solved = '\0';
+
+    set_lec(lec, name);
+    set_method(method);
+    set_sample(from, to);
+    set_comment(comment);
+    set_instruments(instruments);
+    set_block(block);
+    set_tests({ 0.0f });
+    if(date)
+        update_date();
+}
+
+Equation::Equation(const std::string& name, const std::string& lec, const std::string& method, const std::string& from, const std::string& to, 
+    const std::string& comment, const std::string& instruments, const std::string& block, const bool date)
+{
+    this->lec = NULL, 
+    this->clec = NULL, 
+    this->cmt = NULL, 
+    this->instr = NULL, 
+    this->blk = NULL;
+    this->date = 0L;
+    this->solved = '\0';
+
+    set_lec(lec, name);
+    set_method(method);
+    set_sample(from, to);
+    set_comment(comment);
+    set_instruments(instruments);
+    set_block(block);
+    set_tests({ 0.0f });
+    if(date)
+        update_date();
 }
 
-Equation::Equation(const Equation& eq)
+
+Equation::Equation(const Equation& other)
 {
-    c_equation = create_equation_deep_copy(eq.c_equation);
-    endo = eq.endo;
+    copy_from_EQ_obj(static_cast<const EQ*>(&other));
 }
 
 Equation::~Equation()
 {
-    if (c_equation) E_free(c_equation);
+    SW_nfree(this->lec);
+    SW_nfree(this->clec);
```

Comment:
@jmpplan 
Does the function `SW_nfree()` set the passed pointer to `NULL` after it has been freed ?
If not, do you know an easy way to check if a pointer is valid ?

if `SW_nfree()` does not reset the pointer to `NULL`, I can't use:
```c
if(clec != NULL)
{
  ...
}
```
to test if the pointer is valid. Do you know what I mean ?

Reply:
As the argument (here the pointer) is passed by value, it cannot be changed inside `SW_nfree()`. 
To solve that (legitimate) issue, the only way I know is to use a leve l of indirection, for example by storing the allocated references (pointers) in a table and then using references to that table to find the real pointer. BTW, that's one of the reason we had implemented the "Swap" group of functions that does exactly that.

Reply:
Thanks for you answer.

>To solve that (legitimate) issue, the only way I know is to use a leve l of indirection, for example by storing the allocated references (pointers) in a table and then using references to that table to find the real pointer. BTW, that's one of the reason we had implemented the "Swap" group of functions that does exactly that.

That's what C++ `smart pointers` do. 

But in the present case, as I understand your answer, it seems that I have no way to check if a pointer (for example a clec attribute from an instance of EQ) is valid before to call SW_nfree(clec). Am I right ?

To be honest, I am often stuck in the development of my C++ API because I'm trying to free a pointer that has already been freed. Then, to find where that pointer has been freed is very complicated because the C API is very large and some parts are very obscure.